### PR TITLE
Profiles feature branch in Backend

### PR DESCRIPTION
**Pull request by mattias-p**
This is a work in progress for #405.

### CODE REVIEW

**Comment by sandoche2k** [comment]
@mattias-p merge?

**Comment by mattias-p** [comment]
@sandoche2k This branch should to be updated with the implementation first.

**Comment by sandoche2k** [comment]
@mattias-p @matsduf @pamasse please verify the new modifications

**Comment by matsduf** [comment]
The Travis failure should be fixed.

**Comment by mattias-p** [comment]
The TestAgent::new() `config` argument is no longer optional, but the tests have not been updated to reflect this.

File: share/backend_config.ini [hunk 1]
```diff
@@ -16,3 +16,17 @@ interpreter       = perl
 max_zonemaster_execution_time            = 300
 number_of_processes_for_frontend_testing = 20
 number_of_processes_for_batch_testing    = 20
+
+[PUBLIC PROFILES]
+#example_profile_1=/example/directory/test1_profile.json
+#default=/example/directory/default_profile.json
```
Comment (pamasse) [inline share/backend_config.ini]: We must to defined a default public profiles

**Comment by pamasse** [comment]
The main issue with the travis test is that when we install zonemaster-backend we don't set a default public profile

**Comment by mtoma** [comment]
Seems like we have a perl 5.20 specific problem.

Tried to use 'Zonemaster::Backend::RPCAPI'.
#     Error:  Property resolver.defaults.fallback is of type Boolean

The Backend is not directly using the resolver.defaults.fallback option. This seems to be an issue in the Engine.

The same error was fixed in the backend by changing ip_v4/ip_v6 options from Perl value 1 to JSON::PP:true

**Comment by matsduf** [comment]
Travis tests with Perl 5.14 to 5.24 (even versions), but .travis.yml in the develop branch specifies Perl 5.16 to 5.28. Where does Travis get that from? The master branch?

**Comment by vlevigneron** [comment]
@mtoma It comes from : 
https://github.com/zonemaster/zonemaster-engine/blob/develop/lib/Zonemaster/Engine/Profile.pm#L200

This method should accept 0/1/JSON::PP::true/JSON::PP::false, but not false/true (As written line 210, it is something that is quite conceivable).

If it is only on a particular version of Perl, it could be an issue of true/false interpretation in JSON file.
I'll do the check in perl 5.20 and I'll do a patch if needed. Can you create an issue in https://github.com/zonemaster/zonemaster-engine

**Comment by mtoma** [comment]
Issue created:
https://github.com/zonemaster/zonemaster-engine/issues/469

----- Mail original -----
De: "vlevigneron" <notifications@github.com>
À: "zonemaster/zonemaster-backend" <zonemaster-backend@noreply.github.com>
Cc: "Michal TOMA" <michal.toma@afnic.fr>, "Mention" <mention@noreply.github.com>
Envoyé: Jeudi 10 Janvier 2019 14:52:22
Objet: Re: [zonemaster/zonemaster-backend] Profiles feature branch in Backend (#403)

@mtoma It comes from : 
https://github.com/zonemaster/zonemaster-engine/blob/develop/lib/Zonemaster/Engine/Profile.pm#L200

This method should accept 0/1/JSON::PP::true/JSON::PP::false, but not false/true (As written line 210, it is something that is quite conceivable).

If it is only on a particular version of Perl, it could be an issue of true/false interpretation in JSON file.
I'll do the check in perl 5.20 and I'll do a patch if needed. Can you create an issue in https://github.com/zonemaster/zonemaster-engine

**Comment by vlevigneron** [comment]
@mtoma @matsduf It seems it is no longer an issue while We fixed nothing... Travis issue instead ?!?!?
https://travis-ci.org/zonemaster/zonemaster-backend/builds/478313116

**Comment by mtoma** [comment]
Couldn't the message for JSON explain the behaviour?:
https://metacpan.org/release/JSON

Changes for version 4.00 - 2018-12-07
production release
Changes for version 3.99_01 - 2018-12-03
BACKWARD INCOMPATIBILITY: As JSON::XS 4.0 changed its policy and enabled allow_nonref by default, JSON::PP, and thus JSON, also enabled allow_nonref by default
updated backportPP with JSON::PP 3.99_01
allow PERL_JSON_PP_USE_B environmental variable to restore old number detection behavior for compatibility

----- Mail original -----
De: "vlevigneron" <notifications@github.com>
À: "zonemaster/zonemaster-backend" <zonemaster-backend@noreply.github.com>
Cc: "Michal TOMA" <michal.toma@afnic.fr>, "Mention" <mention@noreply.github.com>
Envoyé: Vendredi 11 Janvier 2019 13:31:56
Objet: Re: [zonemaster/zonemaster-backend] Profiles feature branch in Backend (#403)

@mtoma @matsduf It seems it is no longer an issue while We fixed nothing... Travis issue instead ?!?!?
https://travis-ci.org/zonemaster/zonemaster-backend/builds/478313116

**Comment by mtoma** [comment]
I now have this:

Failed test 'use Zonemaster::Backend::RPCAPI;'
#   at t/test_validate_syntax.t line 12.
#     Tried to use 'Zonemaster::Backend::RPCAPI'.
#     Error:  Property net.ipv4 is of type Boolean [TYPE=Bool][FROM=DIRECT][VALUE_TYPE=SCALAR][VALUE=true]
# $VAR1 = bless( do{\(my $o = 1)}, 'JSON::PP::Boolean' );
# Compilation failed in require at /home/travis/build/zonemaster/zonemaster-backend/build_dir/blib/lib/Zonemaster/Engine.pm line 11, <DATA> line 1.
# BEGIN failed--compilation aborted at /home/travis/build/zonemaster/zonemaster-backend/build_dir/blib/lib/Zonemaster/Engine.pm line 11, <DATA> line 1.
# Compilation failed in require at /home/travis/build/zonemaster/zonemaster-backend/build_dir/blib/lib/Zonemaster/Backend/RPCAPI.pm line 19, <DATA> line 1.
# BEGIN failed--compilation aborted at /home/travis/build/zonemaster/zonemaster-backend/build_dir/blib/lib/Zonemaster/Backend/RPCAPI.pm line 19, <DATA> line 1.
# Compilation failed in require at t/test_validate_syntax.t line 12, <DATA> line 1.
# BEGIN failed--compilation aborted at t/test_validate_syntax.t line 12, <DATA> line 1.
Can't locate object method "new" via package "Zonemaster::Backend::RPCAPI" at t/test_validate_syntax.t line 15, <DATA> line 1.

Is that of any help?

----- Mail original -----
De: "vlevigneron" <notifications@github.com>
À: "zonemaster/zonemaster-backend" <zonemaster-backend@noreply.github.com>
Cc: "Michal TOMA" <michal.toma@afnic.fr>, "Mention" <mention@noreply.github.com>
Envoyé: Vendredi 11 Janvier 2019 13:31:56
Objet: Re: [zonemaster/zonemaster-backend] Profiles feature branch in Backend (#403)

@mtoma @matsduf It seems it is no longer an issue while We fixed nothing... Travis issue instead ?!?!?
https://travis-ci.org/zonemaster/zonemaster-backend/builds/478313116

**Comment by vlevigneron** [comment]
@mtoma Perfect, with that I can propose a patch.

**Comment by mtoma** [comment]
Great.
Let me know when I can restart the test.

**Comment by mattias-p** [comment]
I fixed the merge conflict with the develop branch, but I failed to request a review for it. Sorry about that. Please check https://github.com/zonemaster/zonemaster-backend/pull/403/commits/cfcd039f5821b3d2fc6061acbbdec7e11c14a0cc to verify that my resolution is correct.

**Comment by sandoche2k** [comment]
@pamasse @matsduf @mtoma Please verify @mattias-p https://github.com/zonemaster/zonemaster-backend/commit/cfcd039f5821b3d2fc6061acbbdec7e11c14a0cc and approve

**Comment by sandoche2k** [comment]
@mattias-p can we merge?

**Comment by mattias-p** [comment]
@sandoche2k I restarted Travis to verify that that the failing tests pass after zonemaster/zonemaster-engine#472 was merged. I'd also like to rerun Travis after zonemaster/zonemaster-engine#474 is merged to see that it doesn't reintroduce the problemes.

**Comment by mattias-p** [comment]
Perl 5.20 for PostgreSQL and SQLite passed! (no news about MySQL yet but I think it looks promising)

However 5.18 and 5.16 failed for SQLite failed. Pasting the output here (the 5.18 version - 5.16 looks very similar) and restarting them to see if they're reproducible.
```
t/test01.t ................ 1/? Use of uninitialized value in -e at /home/travis/build/zonemaster/zonemaster-backend/build_dir/blib/lib/Zonemaster/Backend/TestAgent.pm line 41.
t/test01.t ................ 11/? Use of uninitialized value in concatenation (.) or string at /home/travis/build/zonemaster/zonemaster-backend/build_dir/blib/lib/Zonemaster/Backend/RPCAPI.pm line 387.
Use of uninitialized value in -e at /home/travis/build/zonemaster/zonemaster-backend/build_dir/blib/lib/Zonemaster/Backend/TestAgent.pm line 41.
Attribute (timestamp) does not pass the type constraint because: Validation failed for 'Num' with value 2,70782804489136 at constructor Zonemaster::Engine::Logger::Entry::new (defined at /home/travis/build/zonemaster/zonemaster-backend/build_dir/blib/lib/Zonemaster/Engine/Logger/Entry.pm line 144) line 79
	Zonemaster::Engine::Logger::Entry::new('Zonemaster::Engine::Logger::Entry', 'HASH(0x4fd1b60)') called at /home/travis/build/zonemaster/zonemaster-backend/build_dir/blib/lib/Zonemaster/Engine/Logger.pm line 24
	Zonemaster::Engine::Logger::add('Zonemaster::Engine::Logger=HASH(0x30f9ea8)', 'RECURSE_QUERY', 'HASH(0x4fc7838)') called at /home/travis/build/zonemaster/zonemaster-backend/build_dir/blib/lib/Zonemaster/Engine/Recursor.pm line 117
	Zonemaster::Engine::Recursor::_recurse('Zonemaster::Engine::Recursor=HASH(0xe84048)', 'Zonemaster::Engine::DNSName=HASH(0x46e2d70)', 'SOA', 'IN', 'HASH(0x46d8498)') called at /home/travis/build/zonemaster/zonemaster-backend/build_dir/blib/lib/Zonemaster/Engine/Recursor.pm line 61
	Zonemaster::Engine::Recursor::parent('Zonemaster::Engine::Recursor=HASH(0xe84048)', 'afnic.fr') called at /home/travis/build/zonemaster/zonemaster-backend/build_dir/blib/lib/Zonemaster/Engine.pm line 134
	Zonemaster::Engine::add_fake_delegation('Zonemaster::Engine', 'afnic.fr', 'HASH(0x46d7d60)') called at /home/travis/build/zonemaster/zonemaster-backend/build_dir/blib/lib/Zonemaster/Backend/TestAgent.pm line 197
	Zonemaster::Backend::TestAgent::add_fake_delegation('Zonemaster::Backend::TestAgent=HASH(0x50607e0)', 'afnic.fr', 'ARRAY(0x50590f0)') called at /home/travis/build/zonemaster/zonemaster-backend/build_dir/blib/lib/Zonemaster/Backend/TestAgent.pm line 133
	Zonemaster::Backend::TestAgent::run('Zonemaster::Backend::TestAgent=HASH(0x50607e0)', 2) called at t/test01.t line 63
	main::run_zonemaster_test_with_backend_API(2) called at t/test01.t line 83
# Tests were run but no plan was declared and done_testing() was not seen.
# Looks like your test exited with 255 just after 22.
t/test01.t ................ Dubious, test returned 255 (wstat 65280, 0xff00)
```

**Comment by mattias-p** [comment]
It seems the 5.16/5.18 problem can occur for other databases too (e.g. PostgreSQL):
* https://travis-ci.org/zonemaster/zonemaster-backend/jobs/478420880
* https://travis-ci.org/zonemaster/zonemaster-backend/jobs/478420881

**Comment by mattias-p** [comment]
Is it a locale problem? The decimal comma looks fishy to me.

**Comment by mattias-p** [comment]
Anyway the timestamp issue (a.k.a the 5.16/5.18 issue) is unaffected by merging of zonemaster/zonemaster-engine#474.

**Pull request merged**